x/liquidstakeibc/keeper/msg_server.go

```diff
@@ -381,6 +391,175 @@ func (k msgServer) LiquidStake(
 	return &types.MsgLiquidStakeResponse{}, nil
 }
 
+// LiquidStakeLSM defines a method for liquid staking tokens using the LSM
+func (k msgServer) LiquidStakeLSM(
+	goCtx context.Context,
+	msg *types.MsgLiquidStakeLSM,
+) (*types.MsgLiquidStakeLSMResponse, error) {
+	ctx := sdktypes.UnwrapSDKContext(goCtx)
+
+	for _, delegation := range msg.Delegations {
+		// check if the ibc denom contains the ibc prefix
+		if !strings.HasPrefix(delegation.Denom, types.IBCPrefix) {
+			return nil,
+				errorsmod.Wrapf(
+					types.ErrInvalidLSMDenom,
+					"IBC denom %s doesn't belong to a LSM token",
+					delegation.Denom,
+				)
+		}
+
+		// parse the ibc denom to extract the original LSM token denom
+		hexHash := delegation.Denom[len(types.IBCPrefix):]
+		hexBytes, err := transfertypes.ParseHexHash(hexHash)
+		if err != nil {
+			return nil, errorsmod.Wrapf(err, "could not parse ibc hash from ibc hex %s", hexHash)
+		}
+
+		// get the denom trace from the parsed ibc denom hex hash
+		denomTrace, found := k.ibcTransferKeeper.GetDenomTrace(ctx, hexBytes)
+		if !found {
+			return nil,
+				errorsmod.Wrapf(
+					types.ErrInvalidLSMDenom,
+					"IBC denom %s doesn't belong to a LSM token",
+					delegation.Denom,
+				)
+		}
+
+		// retrieve the host chain associated with the liquid stake action
+		channelId, found := strings.CutPrefix(denomTrace.Path, fmt.Sprintf("%s/", transfertypes.PortID))
+		if !found {
+			return nil, errorsmod.Wrapf(types.ErrInvalidIBCDenomTrace, "invalid IBC denom trace %s", delegation.Denom)
+		}
+
+		hc, found := k.GetHostChainFromChannelId(ctx, channelId)
+		if !found {
+			return nil, errorsmod.Wrapf(sdkerrors.ErrKeyNotFound, "host chain with channel id %s not registered", channelId)
+		}
+
+		// check if the host chain is active
+		if !hc.Active {
+			return nil, types.ErrHostChainInactive
+		}
+
+		// check if the host chain accepts LSM delegations
+		lsmActiveFlag, found := hc.GetFlag(types.LSMFlag)
+		lsmActive, _ := strconv.ParseBool(lsmActiveFlag)
+		if !lsmActive || !found {
+			return nil, types.ErrLSMNotEnabled
+		}
+
+		// check if the validator is within the module active set
+		operatorAddress, _, found := strings.Cut(denomTrace.BaseDenom, "/")
```

Comment:
i believe lsm module has a function to parse this

```
func (r TokenizeShareRecord) GetShareTokenDenom() string {
	return fmt.Sprintf("%s/%s", strings.ToLower(r.Validator), strconv.Itoa(int(r.Id)))
}
```
i see this is there, but may be the reverse one we can upstream/ create our own function for now

Reply:
that one would need to be executed given a tokenize shared record, which is in the Cosmos chain, I will create methods to parse it ourselves